From 5a73d9a77af737bcf0ddf2e7b8e3ceea4ee18a8a Mon Sep 17 00:00:00 2001
From: Ding Yunliang <yunliang.ding@intel.com>
Date: Fri, 26 Dec 2014 14:31:00 +0800
Subject: [PATCH 186/196] isp: Disable DDR DVFS for camera

It's a know issue for DDR DVFS. If DVFS enabled, ISP timeout observed.
So before camera on, disable DVFS and after camera off, enable it.

Change-Id: Ice72529e469268775c1e83912ea536bec2bcbea2
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-14418
Signed-off-by: Ding Yunliang <yunliang.ding@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/34294
Reviewed-by: Guilin Zhong <guilin.zhong@intel.com>
Reviewed-by: Bingbu Cao <bingbu.cao@intel.com>
Approver: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
---
 .../pci/atomisp2/atomisp_driver/atomisp-regs.h     |  3 +++
 .../pci/atomisp2/atomisp_driver/atomisp_v4l2.c     | 23 ++++++++++++++++++++++
 2 files changed, 26 insertions(+)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp-regs.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp-regs.h
index 0208037..ed9a79f 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp-regs.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp-regs.h
@@ -111,6 +111,9 @@
 #define MRFLD_ISPSSPM0_ISPSSC_MASK     0x3
 #define MRFLD_ISPSSPM0_IUNIT_POWER_ON  0
 #define MRFLD_ISPSSPM0_IUNIT_POWER_OFF 0x3
+#define MRFLD_ISPSSDVFS			0x13F
+#define MRFLD_BIT0			0x0001
+#define MRFLD_BIT1			0x0002
 
 /* MRFLD CSI lane configuration related */
 #define MRFLD_PORT_CONFIG_NUM  8
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c
index a508d74..0b9448b 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c
@@ -356,6 +356,22 @@ done:
 	return 0;
 }
 
+
+/*WA for DDR DVFS enable/disable*/
+void punit_ddr_dvfs_enable(bool enable)
+{
+	int reg = intel_mid_msgbus_read32(PUNIT_PORT, MRFLD_ISPSSDVFS);
+
+	if (enable) {
+		reg &= ~(MRFLD_BIT0 | MRFLD_BIT1);
+	} else {
+		reg |= (MRFLD_BIT0);
+		reg &= ~(MRFLD_BIT1);
+	}
+
+	intel_mid_msgbus_write32(PUNIT_PORT, MRFLD_ISPSSDVFS, reg);
+}
+
 /* Workaround for pmu_nc_set_power_state not ready in MRFLD */
 int atomisp_mrfld_power_down(struct atomisp_device *isp)
 {
@@ -368,6 +384,9 @@ int atomisp_mrfld_power_down(struct atomisp_device *isp)
 	reg_value |= MRFLD_ISPSSPM0_IUNIT_POWER_OFF;
 	intel_mid_msgbus_write32(PUNIT_PORT, MRFLD_ISPSSPM0, reg_value);
 
+	/*WA:Enable DVFS*/
+	punit_ddr_dvfs_enable(true);
+
 	/*
 	 * There should be no iunit access while power-down is
 	 * in progress HW sighting: 4567865
@@ -400,6 +419,10 @@ int atomisp_mrfld_power_up(struct atomisp_device *isp)
 	unsigned long timeout;
 	u32 reg_value;
 
+	/*WA for PUNIT, if DVFS enabled, ISP timeout observed*/
+	punit_ddr_dvfs_enable(false);
+	msleep(20);
+
 	/* writing 0x0 to ISPSSPM0 bit[1:0] to power off the IUNIT */
 	reg_value = intel_mid_msgbus_read32(PUNIT_PORT, MRFLD_ISPSSPM0);
 	reg_value &= ~MRFLD_ISPSSPM0_ISPSSC_MASK;
-- 
1.9.1

